codemp/rd-vulkan/vk_vbo.cpp

```diff
@@ -0,0 +1,1908 @@
+		data = (byte *)ri.Hunk_AllocateTempMemory (dataSize);
+
+		ofsPosition = stride;
+		attr.verts = (vec4_t *)(data + ofsPosition);
+		stride += sizeof (*attr.verts);
+
+		ofsNormals = stride;
+		attr.normals = (vec4_t *)(data + ofsNormals);
+		stride += sizeof (*attr.normals);
+
+		ofsTexcoords = stride;
+		attr.texcoords = (vec2_t *)(data + ofsTexcoords);
+		stride += sizeof (*attr.texcoords);
+
+		ofsBoneRefs = stride;
+		attr.bonerefs = data + ofsBoneRefs;
+		stride += sizeof (*attr.bonerefs) * 4;
+
+		ofsWeights = stride;
+		attr.weights = data + ofsWeights;
+		stride += sizeof (*attr.weights) * 4;
+
+		ofsTangents = stride;
+		attr.tangents = (vec4_t *)(data + ofsTangents);
+		stride += sizeof (*attr.tangents);
+
+		// Fill in the index buffer and compute tangents
+		uint32_t *indices = (uint32_t *)ri.Hunk_AllocateTempMemory(sizeof(uint32_t) * numTriangles * 3);
+		uint32_t *index = indices;
+		vec4_t *tangentsf = (vec4_t *)ri.Hunk_AllocateTempMemory(sizeof(vec4_t) * numVerts);
+
+		surf = (mdxmSurface_t *)((byte *)lod + sizeof(mdxmLOD_t) + (mdxm->numSurfaces * sizeof(mdxmLODSurfOffset_t)));
+
+		for ( n = 0; n < mdxm->numSurfaces; n++ )
+		{
+			mdxmTriangle_t *t = (mdxmTriangle_t *)((byte *)surf + surf->ofsTriangles);
+
+			for ( k = 0; k < surf->numTriangles; k++, index += 3 )
+			{
+				index[0] = t[k].indexes[0] + baseVertexes[n];
+				assert( index[0] >= 0 && index[0] < numVerts );
+
+				index[1] = t[k].indexes[1] + baseVertexes[n];
+				assert( index[1] >= 0 && index[1] < numVerts );
+
+				index[2] = t[k].indexes[2] + baseVertexes[n];
+				assert( index[2] >= 0 && index[2] < numVerts );
+			}
+
+			// Build tangent space
+			//VBO_CalculateTangentsMDXM( surf, tangentsf + baseVertexes[n] );
+
+			surf = (mdxmSurface_t *)((byte *)surf + surf->ofsEnd);
+		}
+
+		assert( index == (indices + numTriangles * 3) );
+
+		surf = (mdxmSurface_t *)((byte *)lod + sizeof (mdxmLOD_t) + (mdxm->numSurfaces * sizeof (mdxmLODSurfOffset_t)));
+
+		for ( n = 0; n < mdxm->numSurfaces; n++ )
+		{
+			// Positions and normals
+			mdxmVertex_t *v = (mdxmVertex_t *)((byte *)surf + surf->ofsVerts);
+			int *boneRef = (int *)((byte *)surf + surf->ofsBoneReferences);
+
+			for ( k = 0; k < surf->numVerts; k++ )
+			{
+				VectorCopy( v[k].vertCoords, *attr.verts );
+				VectorCopy( v[k].normal, *attr.normals );
+
+				attr.verts = (vec4_t *)((byte *)attr.verts + stride);
+				attr.normals = (vec4_t *)((byte *)attr.normals + stride);
+			}
+
+			// Weights
+			for ( k = 0; k < surf->numVerts; k++ )
+			{
+				int numWeights = G2_GetVertWeights(&v[k]);
+				int lastWeight = 255;
+				int lastInfluence = numWeights - 1;
+				for ( w = 0; w < lastInfluence; w++ )
+				{
+					float weight = G2_GetVertBoneWeightNotSlow( &v[k], w );
+					int packedIndex = G2_GetVertBoneIndex( &v[k], w );
+
+					attr.weights[w] = (byte)(weight * 255.0f);		
+					attr.bonerefs[w] = boneRef[packedIndex];
+
+					lastWeight -= attr.weights[w];
+				}
+
+				assert(lastWeight > 0);
```

Comment:
This crashes map loads (in debug), looks like lastWeight == 0 quite frequently, should change to:
`assert(lastWeight >= 0);` ?

Similar issue is also in rend2.

Reply:
Can probably ignore since its just an assert, I will probably modify this to be >= after merge, since I think it was just trying to avoid negatives and I can't see any reason why 0 would be a problem.